Fixes for indentation in multi-mode documents

kevinushey opened the pull request:
NOTE: Let this brew a day or two to think about whether this really is the best solution...

This PR brings a couple of fixes:

It ensures that chunk headers and footers are _not_ tokenized into the underlying code model, e.g.

    - ```{r}
    + foo <- function(x) {
    +     x + 1
    + }
    - ```

Previously, the header / footer bits were entering the code model, and so the indentation logic would get confused (and believe that e.g. there was an un-finished symbol, `` ` ``, hanging around.)

The reindent logic is moved into GWT code. This was necessary for one main reason -- one change in the Ace default mode API that we tried to step over was the fact that the `getNextLineIndent()` prototype changed -- it is now, by default,

    getNextLineIndent(state, line, tab)

rather than

    getNextLineIndent(state, line, tab, tabSize, row)

However, we rely quite heavily on more than just the current line for computing the current indent (we needed that row parameter), so we work around this by fudging the cursor position (so that, in the context of a `getNextLineIndent()` call, we can assume the cursor is at least on that line).

jjallaire commented:
This is definitely a high regression risk PR given how many things the code model impacts (e.g. I know that we had to do some gymnastics to get the code folding behavior we wanted in R markdown and I know what we did was pretty coupled to the code model). If we think this is a good direction then depending on the severity of the issue it resolves we might want to push it out to the next release (I say this mostly because we aren't going to have a huge closedown time after the next preview release as we want to get the new release out soon). Let me know your thoughts.

kevinushey commented:
I feel that it's a pretty important issue to resolve, at least for the final v0.99 release -- basically, indentation (through CMD + I) in R Markdown is broken. That said, this PR needs some more refinement -- I think I should be able to discover a relatively safer fix here and will update this PR + ping you when I have.

jjallaire commented:
In that case yes I agree that it should be in the final release. Let me
know when you've got the revised PR.

On Tue, Mar 24, 2015 at 9:46 PM, Kevin Ushey <notifications@github.com>
wrote:

> I feel that it's a pretty important issue to resolve, at least for the
> final v0.99 release -- basically, indentation (through CMD + I) in R
> Markdown is broken. That said, this PR needs some more refinement -- I
> think I should be able to discover a relatively safer fix here and will
> update this PR + ping you when I have.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/361#issuecomment-85780969>.
>

yihui commented:
Regarding detecting C++ chunks, I plan to support ```` ```{Rcpp} ```` in R Markdown as well in the future. The current syntax ```` ```{r engine='Rcpp'} ```` is a little awkward (https://github.com/yihui/knitr/issues/963).

jjallaire commented:
That's terrific!!! I was hoping we could do something like this. A
couple of other suggestions for ```{Rcpp}:

1. We could automatically #include <Rcpp.h> + using namespace Rcpp

2. We could take all Rcpp chunks within a document and compile them as
a single translation unit (this will be make big difference in
performance when you have 3+ chunks).

Both of these are up for debate as they do have tradeoffs (hidden code
+ no ability to have R code (e.g. setting environment variables)
influence C++ code. Benefits in turn are less boilerplate code and
increased compilation performance.  @eddelbuettel what's your take on
this?

@kevinushey, I think we should go ahead and recognize the {Rcpp}
syntax now so that we are ready for it as soon as it's added to knitr.

On Wed, Mar 25, 2015 at 12:29 PM, Yihui Xie <notifications@github.com> wrote:
> Regarding detecting C++ chunks, I plan to support ```{Rcpp} in R Markdown as
> well in the future. The current syntax ```{r engine='Rcpp'} is a little
> awkward (yihui/knitr#963).
>
> —
> Reply to this email directly or view it on GitHub.

eddelbuettel commented:
:+1: -- most recent Rcpp Gallery piece I wrote started as `.Rmd` but had to be converted to `.cpp` because of the single compilation unit.   I don't yet follow on the 'no R code, env vars' problem.

kevinushey commented:
@jjallaire, I think this should be ready for testing now. Ultimately, the fix was to ensure that code models check that a particular state _ends_ with a particular name. Let me try to summarize the changes I made:

- The code models now check that a particular state _ends_ with a state name, to properly support embedded modes (e.g. we want to recognize `r-cpp-start` within the C++ code model as the `start` state)

The rest is mostly just cleanup + readability related:

- Modes now provide both a 'code begin' and 'code end' pattern; although it turns out these are not strictly required (the way the code folding works, the R code model _must_ see those and hence must know to 'avoid' such headers)

- Outdenting modules now take the underlying code model (since they often need extra information provided in that code model to do the right thing; previous we had them passing a code model as a parameter, which was awkward)

- I implemented some utilities for extracting the mode from a particular state, which I believe makes the code more readable -- e.g. for a state `r-cpp-start`, we extract the current 'mode' as `r-cpp`.

- C++ indentation is now implemented + available for R Markdown documents.

jjallaire commented:
@eddelbuettel the issue is that if you e.g. do a Sys.setenv(PKG_CXXFLAGS = "-fwhatever") within R it won't be picked up by the C++ (because it's compiled in a single pass before the R code executes). I think this is a completely livable constraint though (as we have other ways of doing this via plugins and other attributes we've discussed adding). 

So I'd vote for no #include/using namespace boilerplate required and a single merged translation unit. @yihui whenever you look at implementing this let me know and I can go into more depth on all of this.

yihui commented:
@jjallaire I guess I can spend some time on it later next week. I'll let you know.

jjallaire commented:
@yihui okay, let's chat at whatever point you get to this. I'm hopeful this
will be a pretty straightforward addition.

On Thu, Mar 26, 2015 at 8:28 AM, Yihui Xie <notifications@github.com> wrote:

> @jjallaire <https://github.com/jjallaire> I guess I can spend some time
> on it later next week. I'll let you know.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/361#issuecomment-86569005>.
>
